### PR TITLE
Add better documentation for IResourceDetector and OltpExporter configuration

### LINKED ISSUE #4228: Pass the IServiceProvider in ConfigureResource, WithTracing & WithMetrics
# Feature Request

**Is your feature request related to a problem?**

We need an `IServiceProvider` in `ConfigureResource` and `WithTracing` (and `WithMetrics` if we'd use it) for stuff like the options pattern.

**Describe the solution you'd like:**

A service provider being provided (as an option) to the respective callbacks.

**Describe alternatives you've considered.**

For `ConfigureResource` you could maybe get it done with an extension method, but you're basically rewriting library code at that point. For `WithTracing` I don't see a solution, without accessing private stuff via reflection.

### PR DESCRIPTION
**Pull request by cromefire**
Fixes #4228.

## Changes

- Adds more documentation to make the configurability of the OLTP exporter more clear.
- Better documentation for the `IResourceDetector`

For significant contributions please make sure you have completed the following items:

* [ ] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #4228
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs
Comment (CodeBlanch) [inline src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs:134]: I think we need to remove this. It just won't work for most everything that exists 😢 For example, try this...

`services.AddOpenTelemetry().WithMetrics((builder, sp) => builder.AddOtlpExporter())`

You should get a `NotSupportedException` thrown. The reason is the extensions need the service collection to do options work and it can't be done once the `IServiceProvider` has been created.

There are other existing APIs you can use to accomplish this where it is needed. IMO having this API as a top-level thing will just lead people into trouble.

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs
Comment (CodeBlanch) [inline src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs:185]: Same comment here I think we need to remove this.

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs
Comment (CodeBlanch) [inline src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs:80]: This API I think might work. I want to take it for a spin and try a couple things but I don't see any obvious reasons why it would be trouble to have this.

nit: One thing though, would you reorder the params so it is `IServiceProvider, ResourceBuilder` I think in all our "factory" overloads `IServiceProvider` comes first.

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs
Comment (cromefire) [reply src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs:134]: > There are other existing APIs you can use to accomplish this where it is needed. IMO having this API as a top-level thing will just lead people into trouble.

Sadly not, as to get to services form the DI, you know, you need the DI. So basically I probably have to reimplement the whole Hosting stuff... Yay. Looks like the current architecture sadly doesn't seem to be suitable to that... Our specific use cases needs stuff from the DI to feed into the `AddOtlpExporter` so that seems to be off the table...

Might need to work out something using factories or configure methods and stuff like that, but that'll probably be more complex and would need a full reengineer of the hosting and be therefore unsuitable to this project...

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs
Comment (cromefire) [reply src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs:80]: > nit: One thing though, would you reorder the params so it is IServiceProvider, ResourceBuilder I think in all our "factory" overloads IServiceProvider comes first.

Well, the idea is that it make a lot more sense that the optional parameter comes second so you can just add it, but I mean the ordering doesn't make a huge difference me...

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs
Comment (cromefire) [reply src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs:134]: Also from the looks of it it seems like it should be fine, but sadly your solution doesn't even seem to build currently:
```
CSC: Error CS1705 : Assembly 'OpenTelemetry.Instrumentation.Runtime' with identity 'OpenTelemetry.Instrumentation.Runtime, Version=1.0.0.0, Culture=neutral, PublicKeyToken=7bd6737fe5b67e3c' uses 'OpenTelemetry.Api, Version=1.0.0.0, Culture=neutral, PublicKeyToken=7bd6737fe5b67e3c' which has a higher version than referenced assembly 'OpenTelemetry.Api' with identity 'OpenTelemetry.Api, Version=0.0.0.0, Culture=neutral, PublicKeyToken=7bd6737fe5b67e3c'
```

Comment (cromefire) [reply src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs:134]: So after unloading some projects I got it to build and also found the cuprit, being the need to `ConfigureServices`.

So would it maybe be feasible to add the `IServiceProvider` to the `AddOtlpExporter` configuration. It's kinda not as consistent, but might be actually feasible, as there's a service provider available `AddReader`.

Comment (CodeBlanch) [reply src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs:134]: I'm totally open to making changes! Can you show me what you are trying to do with service provider + AddOtlpExporter? Just curious what the use case is because I haven't seen anyone ask for this yet (that I can recall).

Comment (cromefire) [reply src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs:134]: Well basically because we have a base class throughout a bunch of services, we need to keep stuff pretty generic. For the resources (which should be working already with this PR as you noted) the code would e similar to the following:
```csharp
.ConfigureResource((builder, sp) => {
    var svc = sp.GetRequiredService<Service>();
    builder.AddService("...", instanceId: svc.GetUniqueIdKnowToUs());
})
```

Pretty similar for the tracing:
```csharp
.WithTracing((builder, sp) => {
    var svc = sp.GetRequiredService<Service>();
    builder.AddStuffInstrumentation()
        .AddOtelExporter(options => {
            options.Uri = svc.FetchUriForOtel();
        });
})
```

Comment (cromefire) [reply src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs]: I do have to test, whether this might work:
```csharp
services.AddOptions<OtlpExporterOptions>().Configure<Service>((opts, svc) => {
    // ...
});
```

Comment (CodeBlanch) [reply src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs]: It will, yes! If you just want to configure options, we fully support the Options API (as of 1.4). Just watch out because OtlpExporter has the same options class shared for traces and metrics. If you want to do this style and do individual things for traces & metrics use named options a la:

services.AddOptions<OtlpExporterOptions>().Configure<Service>("traces", (opts, svc) => {... });
services.AddOptions<OtlpExporterOptions>().Configure<Service>("metrics", (opts, svc) => {... });

tracerBuilder.AddOtlpExporter(name: "traces"...);
meterBuilder.AddOtlpExporter(name: "metrics"...);

Comment (CodeBlanch) [reply src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs]: PS: For you resource case check out #4139.

Comment (cromefire) [reply src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs]: Add docs for the exporter case, for the resource case the detector looks nice, but it doesn't seem to be documented at all and I have yet to find a way to like easily register it with Open Telemetry. It doesn't seem like it just picks them up in from the DI or so.

File: src/OpenTelemetry.Extensions.Hosting/README.md
Comment (reyang) [inline src/OpenTelemetry.Extensions.Hosting/README.md:90]: Do we want to use "your" or "my"? (either way works for me as long as we are consistent) https://github.com/open-telemetry/opentelemetry-dotnet/blob/0971088451eb0ec8cef06b852d86cff640e4ec6d/docs/trace/extending-the-sdk/MyExporter.cs#L21

Comment (cromefire) [reply src/OpenTelemetry.Extensions.Hosting/README.md]: I don't care honestly, I always feel more like the documentation is speaking to me, not that the example code is mine, but let's keep it consistent.

File: src/OpenTelemetry/Trace/Builder/TracerProviderBuilderBase.cs
Comment (reyang) [inline src/OpenTelemetry/Trace/Builder/TracerProviderBuilderBase.cs:45]: @CodeBlanch any suggestion how to improve the wording here? e.g. "what happened; what to do"

**Comment by reyang** [comment]
@cromefire would you update the PR title please? (this PR is exposing new API, which is more than "add better documentation")

File: src/OpenTelemetry/Trace/Builder/TracerProviderBuilderBase.cs
Comment (cromefire) [reply src/OpenTelemetry/Trace/Builder/TracerProviderBuilderBase.cs:45]: I didn't even mean to touch that, I think the rider formatter just did a little too much formatting, let me revert that.

**Comment by cromefire** [comment]
Also added #4139 to the fixes list.

**Review by cijothomas — DISMISSED** [review]
Blocking temporarily to prevent accidental merges, as we are in the process of stable release. Will unblock once the release process is done.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/README.md
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/README.md:77]: @cromefire Hey would you be open to moving this into its own PR? We can get this in more quickly than the new public API changes.

Comment (cromefire) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/README.md:77]: Sure, I'll move it in the coming days.

Comment (cromefire) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/README.md:77]: Converted this PR

**Comment by cromefire** [comment]
Made this into the PR for the documentation

File: src/OpenTelemetry.Extensions.Hosting/README.md
Comment (cijothomas) [inline src/OpenTelemetry.Extensions.Hosting/README.md:105]: add Metrics too, to make it very clear to users that the resource is applied to traces and metrics both.

Comment (CodeBlanch) [reply src/OpenTelemetry.Extensions.Hosting/README.md:105]: nit: You should be able to remove the `!` on the `sp` param because it is no longer nullable:

Before: `sp!.GetRequiredService<MyResourceDetector>()`
Updated: `sp.GetRequiredService<MyResourceDetector>()`

Comment (cromefire) [reply src/OpenTelemetry.Extensions.Hosting/README.md]: Done

**Comment by cromefire** [comment]
Just waiting for merge, bot

File: src/OpenTelemetry.Extensions.Hosting/README.md
Comment (CodeBlanch) [inline src/OpenTelemetry.Extensions.Hosting/README.md:102]: What if we made this slightly more interesting by showing something being injected?

```csharp
public class MyResourceDetector : IResourceDetector
{
    private readonly IWebHostEnvironment webHostEnvironment;

    public MyResourceDetector(IWebHostEnvironment webHostEnvironment)
    {
        this.webHostEnvironment = webHostEnvironment;
    }

    public Resource Detect()
    {
        return ResourceBuilder.CreateEmpty()
            .AddService(serviceName: this.webHostEnvironment.ApplicationName)
            .AddAttributes(new Dictionary<string, object> { ["host.environment"] = this.webHostEnvironment.EnvironmentName })
            .Build();
    }
}
```

Comment (cromefire) [reply src/OpenTelemetry.Extensions.Hosting/README.md:102]: The environment is injected, I can add something else too if you want.

Comment (CodeBlanch) [reply src/OpenTelemetry.Extensions.Hosting/README.md:102]: @cromefire I don't see any changes on the diff did you forget to push maybe? 😄

Comment (cromefire) [reply src/OpenTelemetry.Extensions.Hosting/README.md:102]: No I'm just too stupid to use the GitHub app... Will fix

Comment (cromefire) [reply src/OpenTelemetry.Extensions.Hosting/README.md]: Fixed

**Review by CodeBlanch — APPROVED** [review]
LGTM

**Pull request merged**
